### PR TITLE
Implement limit active tasks queue

### PR DESCRIPTION
**Pull request by aledeganopix4d**
# Why do we need this PR?
Currently when using the container placement strategy `limit-active-tasks` when all the workers are busy waiting for a free worker the first task to be executed after a worker is free again will be picked randomly (e.g. whichever task queries the worker pool first after the worker has been freed).
This creates an unfair queue where tasks could be executed after tasks that joined the queue later, moreover, if the queue is sufficiently big, a task could be waiting a lot of time before being executed, just by the nature of the stochastic process of tasks execution.

As a bonus effect, with this PR we can identify uniquely tasks that are in queue and thus exporting a metric on how long are the different queues based on `platform`, `team workers`, `tags`.

# Changes proposed in this pull request
* Add a DB table to represent the tasks queue
* When `limit-active-tasks` is enabled, tasks join a queue according to the platform, dedicated team workers (if any), task tags (if defined)
* The task wait in queue and report the progress in the UI until it reaches the position 1 (e.g. the front of the queue), at that point will still wait for the first available worker
* Export a new metric with the queues length, can be useful to autoscale workers!

# Contributor Checklist
- [X] Unit tests
- [ ] Integration tests (if applicable)
- [ ] Updated documentation (located at https://github.com/concourse/docs)
- [ ] Updated release notes (located at https://github.com/concourse/concourse/tree/master/release-notes)

# Reviewer Checklist
> This section is intended for the core maintainers only, to track review progress. Please do not
> fill out this section.
- [ ] Code reviewed
- [ ] Tests reviewed
- [ ] Documentation reviewed
- [ ] Release notes reviewed
- [ ] PR acceptance performed

### CODE REVIEW

File: atc/worker/client.go [hunk 1]
```diff
@@ -285,6 +316,52 @@ func (client *client) chooseTaskWorker(
...
+				if dedicatedWorker {
+					teamId = workerSpec.TeamID
+				} else {
+					teamId = 0
+				}
+				tags := containerSpec.Tags
+				sort.Strings(tags)
+				workerTags := strings.Join(tags, "_")
+				queue_position, qlen, err := client.tasksQueue.FindOrAppend(taskId, workerSpec.Platform, teamId, workerTags, logger)
+				defer client.tasksQueue.Leave(taskId, logger)
```
[the remark this answers is not in the document]
Comment (aledeganopix4d) [reply atc/worker/client.go]: 1. I want to keep the full logic in the TasksQueue and not in Client. This is similar, for instance, to `pool.FindOrChooseWorkerForContainer`
2. It's called multiple times and it's idempotent. This saves the complexity that would be required just to know if the task is being queued on this cycle or it was before.

File: atc/db/tasks_queue.go [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+	"database/sql"
+	"fmt"
+
+	"code.cloudfoundry.org/lager"
+	sq "github.com/Masterminds/squirrel"
+)
+
+//go:generate counterfeiter . TasksQueue
+
+type TasksQueue interface {
```
[the remark this answers is not in the document]
Comment (aledeganopix4d) [reply atc/db/tasks_queue.go]: Renamed as suggested.

File: atc/db/tasks_queue.go [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+	FindOrAppend(string, string, int, string, lager.Logger) (int, int, error)
+	FindQueue(string) (string, int, string, error)
+	Leave(string, lager.Logger)
```
[the remark this answers is not in the document]
Comment (aledeganopix4d) [reply atc/db/tasks_queue.go]: Rename to `Dequeue`.

File: atc/worker/client.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (marco-m) [reply atc/worker/client.go]: >  2\. It's called multiple times and it's idempotent. This saves the complexity that would be required just to know if the task is being queued on this cycle or it was before.

It might be idempotent from the point of view of the side-effects on the database.

The problem is on the Go runtime. What I think will happen is that the `defer`s will add up to the runtime defer linked list, each time we enter this code path. Depending on how often this code path is taken for the same queue element, this can have a serious impact on the runtime resources and behavior.

[the remark this answers is not in the document]
Comment (aledeganopix4d) [reply atc/worker/client.go]: After giving some thought I implemented an easy solution to avoid calling the `defer` more than once.
Thank you for the feedback.

File: atc/db/task_queue.go [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+	}
+}
+
+func (queue *taskQueue) FindOrAppend(id string, platform string, team_id int, worker_tag string, logger lager.Logger) (int, int, error) {
+	// Returns the position and the total queue length for a given id
+	var (
+		position int
+		len      int
+	)
+	ex_platform, ex_team_id, ex_worker_tag, err := queue.FindQueue(id)
```
Comment (taylorsilva) [inline atc/db/task_queue.go]: The code base (and golang in general) uses camel case for go variable names. Snake case gets used for database column names and queries.

```suggestion
	exPlatform, exTeamId, exWorkerTag, err := queue.FindQueue(id)
```

File: atc/db/task_queue.go [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+		conn: conn,
+	}
+}
+
+func (queue *taskQueue) FindOrAppend(id string, platform string, team_id int, worker_tag string, logger lager.Logger) (int, int, error) {
+	// Returns the position and the total queue length for a given id
+	var (
+		position int
+		len      int
+	)
```
[the remark this answers is not in the document]
Comment (marco-m) [reply atc/db/task_queue.go]: Named return variables by themselves: why not. The problem is the naked return, which makes the code error-prone and less readable than explicit return.

**Comment by ddadlani** [comment]
Hi @aledeganopix4d, thanks for submitting this PR! 

The idea of adding queues for scheduling is great, and is something we've been discussing for a while (#3695 for example). However, we feel that it makes more sense to implement queuing regardless of the placement strategy used, since it could improve the performance of all of them.

Would you be open to implementing this queue system for all placement strategies, and not just `limit-active-tasks`?

The process would involve writing an architecture RFC to flesh out the design decisions to be made and solicit community feedback, and then to implement it according to those changes. We can set up a simple RFC for you, give you any context required and help you along the way as needed.

Let us know if you are open to working on this.

/cc @vito, @deniseyu , @kcmannem, @xtreme-sameer-vohra

File: atc/db/task_queue.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (aledeganopix4d) [reply atc/db/task_queue.go]: I implemented named return variables in all the functions of this interface (but keeping the returns explicit).
I also change `len` to `length` at this particular line.

File: atc/db/task_queue.go (hunk 4 above)
Comment (aledeganopix4d) [reply atc/db/task_queue.go]: I've replaced all snakes to camels :) :snake: :arrow_right: :camel:

**Comment by aledeganopix4d** [comment]
> Hi @aledeganopix4d, thanks for submitting this PR!
> 
> The idea of adding queues for scheduling is great, and is something we've been discussing for a while (#3695 for example). However, we feel that it makes more sense to implement queuing regardless of the placement strategy used, since it could improve the performance of all of them.
> 
> Would you be open to implementing this queue system for all placement strategies, and not just `limit-active-tasks`?
> 
> The process would involve writing an architecture RFC to flesh out the design decisions to be made and solicit community feedback, and then to implement it according to those changes. We can set up a simple RFC for you, give you any context required and help you along the way as needed.
> 
> Let us know if you are open to working on this.
> 
> /cc @vito, @deniseyu , @kcmannem, @xtreme-sameer-vohra

Hello @ddadlani and thank you for your feedback.

I agree that's a great idea to implement some form of queuing across all placement strategies but that's quite a different problem with respect to the one I'm trying to solve here :)
For instance I have a really hard time imagining how the queue should behave for `volume-locality` and as you say it would require heavy community involvement.

On the other hand I'm confident that the queuing method I'm proposing in this PR is sound enough to test in production where we could also get some operational feedback and quite likely some idea on how to implement queues on a more general level.

The purpose of this PR, moreover, is not really performance related: `limit-active-tasks` already guarantees a certain level of that, this PR has a more "social" scope: ensure that no one task remains in line "forever" when the `limit-active-tasks` is chosen (alongside a finite number of active tasks per worker).

**Comment by vito** [comment]
I echo @ddadlani's concern that this is extending what was supposed to be a stop-gap solution. :slightly_smiling_face: 

In my mind a queue implementation should be applicable regardless of the placement logic. We don't need one implemented for each. The current `limit-active-tasks` is basically just `fewest-build-containers` but there's no reason there couldn't be a "place based on volume locality but only on workers with < N active tasks" right?

Can the logic for whether a worker should be used *at all* be decoupled from the logic determining which *usable* workers we should place a container on? If we do that, I think we can eliminate `limit-active-tasks` entirely and compose these rules instead.

And just a quick sidebar on queueing: we should be queueing the *placement*, not the *creation* of containers - i.e. something should be placed-and-created at the same time. Otherwise you could have containers "queued" on to workers that don't free up as quickly as other workers. Does that make sense? :thinking:

**Comment by aledeganopix4d** [comment]
@vito, @ddadlani 
I disagree that it's possible to have a 

> queue implementation should be applicable regardless of the placement logic

**unless** we change both the implementation and (more importantly) the semantic of "placement strategy".
Let me illustrate why:
`volume-locality` and `random` as they are implemented now they contrast with the very idea of the queue: a task (or even) a step would be waiting in queue for its turn to be executed, just then to be assigned to an arbitrary worker chosen for a reason unrelated to the queue system in the first place.

What we could do, then, is `prioritize this worker if possible, chose another one otherwise`. But then this is no longer a "placement" strategy but an "affinity" and thus both the implementation and the operator expectation would be drastically different to the current status.

At this point we would be talking on how to implement `affinity` and `requirements` (e.g. as they are used in K8S and Nomad) in Concourse and I don't think that can be done without re-implementing from scratch the scheduling.

This PR implements 2 very important feature (for my team) on an existing placement strategy:
- Execution fairness
- Tasks backlog size

The above two features wouldn't even make sense in the other existing placement strategy as they are today.

Finally:
> And just a quick sidebar on queueing: we should be queueing the placement, not the creation of containers - i.e. something should be placed-and-created at the same time. 

But this is how it is implemented here because the queue blocks `client.pool.FindOrChooseWorkerForContainer` which, as far as I understand, finds the worker and then place the container on it.

**Comment by vito** [comment]
Hmm I'm not sure I follow. :thinking: Maybe I just have different expectations out of scheduling/placement. Let me try to write out my expected mechanics:

Let's say...:

* I have 3 workers, w1 w2 and w3.
* I'm limiting the active build containers on a worker to 10.
* I'm using `volume-locality` placement strategy.
* I'm trying to place a container for which w1 has 1 local volume, w2 has 2 local volumes, and w3 has 3 local volumes.

Given this worker container count:

| w1 | w2 | w3 |
| - | - | - |
| 2 | 6 | 8 |

I would expect the container to be placed on **w3**.

Given this worker layout:

| w1 | w2 | w3 |
| - | - | - |
| 2 | 6 | 10 |

I would expect the container to be placed on **w2**.

And given this worker layout:

| w1 | w2 | w3 |
| - | - | - |
| 10 | 10 | 10 |

I would expect my container placement to be "queued" until a worker is available.

I would rather have my tasks schedule as soon as they can than wait for the strictly most optimal worker. This is good enough for me as I don't expect my workers to be "at capacity" all the time, so overall the placement would still be a net benefit. Placement is inherently an optimization, and I expect the system to make a best effort at placing things in the optimal location, but *also* start running things as soon as possible - otherwise we're just trading time spent on a suboptimal worker for time spent waiting for the optimal worker.

This seems like something that could apply to any placement logic. Is your expected behavior different?

> But this is how it is implemented here because the queue blocks client.pool.FindOrChooseWorkerForContainer which, as far as I understand, finds the worker and then place the container on it.

Sorry, to clarify I wasn't calling that out as a correction, it was just a subtle thing I wanted to bring up to make sure it was covered since I thought of it while I was writing my comment. :slightly_smiling_face: Sounds like it is.

**Comment by aledeganopix4d** [comment]
@vito, what you suggest is possible but it would require great efforts not only from a software development perspective, but especially from an architectural direction and inter-teams agreement perspective.

Regarding
> - I'm limiting the active build containers on a worker to 10.
> - I'm using volume-locality placement strategy.

With the current implementation this is not possible for any placement strategy because they have no idea of “active build containers”.
So, very broadly speaking, to implement what you and @ddadlani are suggesting we would require to:

1. Merge the `limit-active-tasks` into all the other strategies, so they can be aware of “busy” workers
2. Implement a global task queue (or even step queue as @ddadlani was suggesting) in front of any worker selection 
3. Rethink which placement strategies still make sense after the rework (e.g. I don’t think that `fewest-build-containers` would still make a lot of sense after that)

And this is only the effort on the “coding” side, because on the community side there would still be a lot to do:
1. Define an RFC to refactor the entire scheduling of Concourse
2. Reach an agreement with the external community and the internal Pivotal teams
3. Coordinate the refactoring with the active work performed at Pivotal to ensure no one step on each other toes
4. Fully implement functional/end-to-end testing as such refactoring would need a really thorough check

For the above reasons I, and my team, believe that the above is an effort outside of our capabilities, even more so, outside of any external contributor capabilities (especially considering how many outstanding issues are still open around the scheduling topic).

This PR stems from two very precise (and in our experience important) requirements:
1. Ensure that tasks are not blocked in the queue for a long time just by the random nature of the tasks execution
2. Monitor the length of the queues to react in real time to them

We do understand your resistance to introduce even more ad-hoc code in `limit-active-tasks` and I would be happy to work on a middle ground solution. Unfortunately I cannot think how to implement the above requirements otherwise.
If you can think of a reasonable compromise I would be glad to consider it.

**Comment by sambarluc** [comment]
I was pointed to this PR by my company's CI team. They were trying to explain me why my 5 minutes task took 50 minutes to run (it took 45 minutes to get a free worker, even if workers were not overloaded).

I know less then nothing about Concourse internals, and never wrote a line of Go. However, purely as a user, I must say that the Concourse scheduling system (or, should I say, its absence?) is a source of constant pain, in an otherwise remarkably powerful and flexible tool that I use every day. In an environment where builds and tests require substantial resources, and where expensive workers are thus relatively scarce, I think that a "fair" (using @aledeganopix4d's term) queueing system is badly needed. Furthermore, I don't see the point of delaying this, in favour of what sounds like an overall restructuring of the entire Concourse scheduling system, that probably requires extensive coordination and testing from multiple teams.

TL;DR: I need it, please get on with it!

**Comment by vito** [comment]
All I'm asking for is a bit more collaboration. :sweat_smile: The reality of the situation is that there are many users of Concourse and only a few people responsible for maintaining it so we have to be more careful about its architecture and design. Long-term health of the project is more important than the immediate needs of any particular contributor or company.

@aledeganopix4d No one's asking for a full refactor of the Concourse scheduler. All I'm asking is for us to reach an agreement on what we want in the long term, as it was made clear long before this PR that `limit-active-tasks` is a short-term solution in the absence of a proper queue. Now that we're adding a proper queue, it seems wrong to add it to the `limit-active-tasks` placement strategy - it was supposed to be a stopgap, not the end solution!

So far y'all have proven capable of writing significant amounts of code and engaging with us on GitHub (and even getting other people from your company involved, judging by the comments and upvotes, lol). I promise that these are the very same capabilities necessary to do this the "right way" - the only difference is I would rather have conversations happen *first* rather than have y'all do a lot of upfront work and just pressure us to merge it in. Having to coordinate is a fact of life, and is not something to avoid. It's why we have the RFC process, and it's why we try to make as many decisions as possible in public on GitHub.

As for the technical strategy, saying this would be a complete overhaul of the Concourse scheduler is a bit overblown. Workers already keep track of active build containers; it's how the `fewest-build-containers` strategy works. To me there's a pretty clear-cut path from here to a better long-term design. Most of the work is already done, it just needs to be moved around a bit - and no, I don't see that being a complete overhaul.

I'm trying to get to a point where we can both see that, but so far all of our suggestions and concerns are being deflected in favor of just merging this in. This is frustrating for us because we're heavily invested in Concourse's sustainability and success and we're the ones responsible for maintaining and supporting it.

We've already done the short-term change by merging the stopgap solution in the first place. I only ask that you be more receptive to our concerns about the long-term design, and help us reach an agreement that the community can carry forward together.

If you want to get scheduling to the finish line, we need to get past the stopgap solution. I would really like to collaborate with y'all on this. We're happy to help in any way possible. For example, we could kick off an RFC clearly outlining what I'm trying to propose. If that RFC really *does* prove to be beyond your capabiltiies - which I truly doubt based on what I've seen so far - we can maybe even fit it into our own roadmap, but at that point I can't make many promises about when it would be delivered. (Runtime already has a ton of high priority work.)

**Comment by ddadlani** [comment]
Hi @aledeganopix4d, in addition to what @vito said, I am also leaving Concourse and as such, there is a decent amount of restructuring going on. @xtreme-sameer-vohra, @kcmannem , @deniseyu and of course @vito will be able to help out with the RFC process and implementation of the general step queue, but there will be a few delays until the team adjusts to the new flow.

**Comment by aledeganopix4d** [comment]
Dear all,
thank you for the feedback.
After some discussion within the team we decided that we will write the RFC proposal and make sure that an agreement will be met.

Meanwhile we decided that this code was still valuable for our developers and thus decided to maintain it in a public fork, for as long as it takes to implement said RFC.
Should anyone reading this be interested the release is here: https://github.com/Pix4D/concourse/releases
(this is what currently we have deployed in prod)

I'm going to close this PR now.

**Comment by vito** [comment]
@aledeganopix4d Thank you!

**Comment by marco-m-pix4d** [comment]
For completeness, the (abandoned) RFC mentioned in this thread is https://github.com/concourse/rfcs/pull/43
